File: kth-smallest-element-in-a-bst/kth-smallest-element-in-a-bst.py
#Time: O(N)
#Spce: O(N)
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: TreeNode, k: int) -> int:
        
        def inorder(root):
            if root:
                inorder(root.left)
                temp.append(root.val)
                inorder(root.right)
        
        temp = []
        inorder(root)
        return temp[k-1]

File: kth-smallest-element-in-a-bst/kth-smallest-element-in-a-bst.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
     Stack<TreeNode> stack = new Stack<TreeNode>();
    public int KthSmallest(TreeNode root, int k) {
        Helper(root);
        TreeNode n = null;
        
        while(k>0)
        {
            n = stack.Pop();
            k--;
            if(k==0) break;
            Helper(n.right);
        }
        return n.val;
      
    }
    
    private void Helper(TreeNode node)
    {
        while(node!=null)
        {
            stack.Push(node);
            node = node.left;
        }
    }
}
